### ISSUE TITLE
Question regarding -s parameter (similarity threshold for duplicate haplotigs)

### ISSUE THREAD

**Issue by pjm43** [chhylp123/hifiasm#218]
I'm assembling a highly heterozygous plant genome with HiC data using the defaults (0.16.1-r375).  The haplotype assemblies are very different sizes (hap1_p_ctg=4.55 Gb; hap2_p_ctg=243 Mb; hic.p_ctg=4.36 Gb).  From the README the solution for this situation seems to be decrease -s and to verify the --hom-cov.  I'm trying to 1) figure out how much I should decrease the -s parameter and 2) verify that the --hom-cov was correctly called. The slurm output indicated that:

[M::purge_dups] homozygous read coverage threshold: 57

Which appears very close to the homozygous peak in the slurm output: [slurm-45588501.out.txt](https://github.com/chhylp123/hifiasm/files/7695785/slurm-45588501.out.txt).  Perhaps it could be set to 58, but I'm guessing that would make very little difference - would you agree?

Thus, I think what I really need to do is decrease -s.  Can you provide any guidance on how much to reduce it?  I've seen some posts where they decrease to 0.10 - I'm just interested in what your suggestion/guidance would be in my situation.

Thanks in advance for the guidance!

**Comment by chhylp123** [comment]
It seems the heterozygosity rate of your sample is very high, so that hifiasm missed most homologous pairs. `-s` is set case by case but for your case it should be set very small. Maybe you can try different `-s` like `-s0.3`, `-s0.2` or `-s0.1` to see if two haplotypes are balanced. These procedures should be fast as they can reuse the bin files.

**Comment by pjm43** [comment]
Thanks for you the very fast response!  I realize that I can simply restart using -o set to the exact same prefix as the previous run - will this overwrite the previous data?  Can I launch multiple runs in parallel with different -s values or do I need to do this in serial?  Lastly, do you agree that hifiasm correctly picked --hom-cov?

Thanks again!

**Comment by chhylp123** [comment]
It will overwrite previous data. If you want to launch multiple runs in parallel, you can create soft links to the bin files, and run hifiasm with soft lines. And I agree `--hom-cov` is correct.

**Comment by pjm43** [comment]
Thanks again for the quick replies.  Much appreciated.

**Comment by pjm43** [comment]
I apologize for reopening this thread.  I have rerun the assembly with different -s values.  None, however, produce completely unzipped haplotypes:

<img width="928" alt="image" src="https://user-images.githubusercontent.com/22619807/145685080-8e39cd15-b306-4932-830b-d38dc71b8c29.png">

The discrepancy in size decreases between the two haplotype assemblies, but neither reaches the expected genome size (2.4 Gb) even at -s 0.05.  Are there any other parameters that I might try to reach a balance between the two haplotypes?  I thought about going below 0.05 for -s, but you'll notice that the balance actually starts to gets worse below 0.10 again (i.e., the balance gets better between the two haplotypes until you reach 0.10 at which point it starts to get more unbalanced).  Would you suggest that keep going lower for -s or try a different parameter?

Any help would be appreciated.  Thanks.

**Comment by chhylp123** [comment]
It is very abnormal. Would you mind sharing the bin files with us? I have no idea what happened.

**Comment by chhylp123** [comment]
Also, could you please try https://github.com/chhylp123/hifiasm/releases/tag/0.16.0? And what are the size of `bp.hap1` and `bp.hap2` without Hi-C?

**Comment by pjm43** [comment]
The bin files are quite large:  

<img width="742" alt="image" src="https://user-images.githubusercontent.com/22619807/145687377-3ed9b603-17c0-46ce-b451-bd301817bee5.png">

I could upload them to a ftp server to give you access to them (or if you have a place you would prefer that I upload them I could do that as well.  Just let me know.

I will download and rerun with the hifiasm version you suggest (r369), although I think it an older version than the one I'm using (r375)?  Should I rerun with the defaults or set -s to something different?   I'll also rerun with hifiasm r375 now without the HiC data and report back what the bp.hap1 and bp.hap2 are without the HiC data.

**Comment by chhylp123** [comment]
Probably no need to run 0.16.0 (r369). I really appreciate if you can let me know the size of `bp.hap1` and `bp.hap2` with default parameters. Could you please upload the bin files to an ftp server? I will download them for debugging. Thank you so much.

**Comment by chhylp123** [comment]
By the way, it would be good to make sure if both Hi-C and HiFi come from the sample.

**Comment by pjm43** [comment]
Yes, I will upload the files to our ftp server and then give you credentials for you to down load.  Can you share your email with me to give you the credentials confidentially (email me at Jeff_Maughan@byu.edu)?  

I'm running the NoHiC assembly now and will share those results when I get them.

I'll try to verify that the Hifi sequence and the HiC data are from the same sample (I'll need to talk to my collaborators to verify on Monday).  

Thanks again for your help.

**Comment by chhylp123** [comment]
My email is: hcheng@jimmy.harvard.edu. Thank you so much.

**Comment by caiobrunharo** [comment]
@pjm43 Did you solve this issue of having very different haplotype sizes? I am also assembly the genome of a plant species with high heterozygosity rates and am having similar issues. I also tried different -s, and adjusted the --hom-cov and --purge-cov parameters but it didn't reduce the different in sizes. If you could share any insights I would really appreciate it. Thanks!

**Comment by chhylp123** [comment]
This issue is caused by wrong Hi-C reads. To quickly check if it is the problem of Hi-C, you should have a look at the size of `bp.hap1` and `bp.hap2` without Hi-C.

**Comment by pjm43** [comment]
@caiobrunharo.  I don't have a definitive answer to our problem, yet.  But I think I know what happened - it wasn't that we had the wrong Hi-C reads, only that we didn't have the full set of reads.  Upon inspection of the fastq files for the Hi-C data, I realized (belatedly) that what we were sent were the QC set of Hi-C reads, which is a very small portion of the total read set.  We are now waiting to get the full set.  Once we have that, I'm hopeful our issue will be resolved.

**Comment by pjm43** [comment]
I apologize for reopening this thread, but I was able to get the correct Hi-C and have since run it.  I'm attaching the hifiasm output here: [slurm-46199993.out.txt](https://github.com/chhylp123/hifiasm/files/7953408/slurm-46199993.out.txt). The assembly was run with default values (-s 0.55).  The hap1 and hap2 were not equal in size (4.0 Gb and 3.7 Gb) and the genome size was much too large and there was too much duplication in the hap assemblies, hence I re-ran the assembly again at various -s values:

<img width="422" alt="image" src="https://user-images.githubusercontent.com/22619807/151434095-67ac412d-f62b-4ff1-83c1-25a5308e56e9.png">

Using different -s values die equalize the haplotig files and began to reduce the genome size/duplication we were seeing.  We anticipate a haplotype assembly to be ~2.5 Gb.  Indeed we started to approach the expected genome size as we drop the -s value, but we are concerned that the duplication (as measured by BUSCO) is still too high, even at -s 0.05.  Any thoughts or advice about how to approach this assembly to further reduce the *duplication* as well as continue to reduce the *genome size*?  Would continuing to pushing -s down (to 0.01) be wise or would you suggest another approach?

Thank you for you advice,

Jeff

**Comment by chhylp123** [comment]
Could you please show the homozygous coverage threshold? I was wondering the issue of size is mostly caused by wrong homozygous coverage threshold, instead of `-s` (see FAQ: https://hifiasm.readthedocs.io/en/latest/faq.html#for-hi-c-integrated-assembly-why-the-assembly-size-of-both-haplotypes-are-much-larger-than-the-estimated-genome-size).

**Comment by pjm43** [comment]
These are the coverage thresholds (as pulled from the slurm):
[M::purge_dups] homozygous read coverage threshold: 57
[M::purge_dups] purge duplication coverage threshold: 72
[M::stat] # heterozygous bases: 4945418176; # homozygous bases: 62118749

Is this what you were looking for?
Jeff

**Comment by chhylp123** [comment]
Thanks. Looks like OK. So what's the homozygous coverage of your dataset?

**Comment by pjm43** [comment]
I think from looking at the slurm that the homozygous peak is near 57 - do you agree (see below)?  I will note that the genome is highly heterozygous (diploid, outcrossing plant).

```
[M::ha_analyze_count] highest: count[28] = 3182687
[M::ha_hist_line]     1: ****************************************************************************************************> 5363731
[M::ha_hist_line]     2: ******************* 601341
[M::ha_hist_line]     3: ***** 173902
[M::ha_hist_line]     4: *** 89551
[M::ha_hist_line]     5: ** 67559
[M::ha_hist_line]     6: ** 63200
[M::ha_hist_line]     7: ** 70246
[M::ha_hist_line]     8: *** 82145
[M::ha_hist_line]     9: *** 101916
[M::ha_hist_line]    10: **** 130497
[M::ha_hist_line]    11: ***** 162089
[M::ha_hist_line]    12: ****** 204303
[M::ha_hist_line]    13: ******** 257642
[M::ha_hist_line]    14: ********** 324962
[M::ha_hist_line]    15: ************* 407719
[M::ha_hist_line]    16: **************** 517693
[M::ha_hist_line]    17: ********************* 668579
[M::ha_hist_line]    18: *************************** 850105
[M::ha_hist_line]    19: ********************************** 1069557
[M::ha_hist_line]    20: ****************************************** 1333584
[M::ha_hist_line]    21: *************************************************** 1620191
[M::ha_hist_line]    22: ************************************************************* 1927942
[M::ha_hist_line]    23: ********************************************************************** 2235543
[M::ha_hist_line]    24: ******************************************************************************* 2528773
[M::ha_hist_line]    25: **************************************************************************************** 2785910
[M::ha_hist_line]    26: ********************************************************************************************** 2984430
[M::ha_hist_line]    27: ************************************************************************************************** 3121979
[M::ha_hist_line]    28: **************************************************************************************************** 3182687
[M::ha_hist_line]    29: *************************************************************************************************** 3146257
[M::ha_hist_line]    30: *********************************************************************************************** 3035598
[M::ha_hist_line]    31: ****************************************************************************************** 2859068
[M::ha_hist_line]    32: ********************************************************************************** 2621645
[M::ha_hist_line]    33: ************************************************************************** 2349358
[M::ha_hist_line]    34: ***************************************************************** 2061201
[M::ha_hist_line]    35: ******************************************************** 1768518
[M::ha_hist_line]    36: *********************************************** 1487502
[M::ha_hist_line]    37: *************************************** 1237428
[M::ha_hist_line]    38: ******************************** 1016826
[M::ha_hist_line]    39: ************************** 823249
[M::ha_hist_line]    40: ********************* 668494
[M::ha_hist_line]    41: ***************** 546329
[M::ha_hist_line]    42: ************** 455032
[M::ha_hist_line]    43: ************ 394187
[M::ha_hist_line]    44: *********** 357628
[M::ha_hist_line]    45: *********** 342129
[M::ha_hist_line]    46: *********** 340841
[M::ha_hist_line]    47: *********** 348778
[M::ha_hist_line]    48: ************ 369591
[M::ha_hist_line]    49: ************ 386970
[M::ha_hist_line]    50: ************* 412782
[M::ha_hist_line]    51: ************** 437329
[M::ha_hist_line]    52: *************** 462807
[M::ha_hist_line]    53: *************** 481642
[M::ha_hist_line]    54: **************** 502879
[M::ha_hist_line]    55: **************** 520820
[M::ha_hist_line]    56: ***************** 530923
[M::ha_hist_line]    57: ***************** 532466
[M::ha_hist_line]    58: ***************** 534343
[M::ha_hist_line]    59: **************** 522610
[M::ha_hist_line]    60: **************** 508699
[M::ha_hist_line]    61: *************** 492083
[M::ha_hist_line]    62: *************** 469400
[M::ha_hist_line]    63: ************** 441366
[M::ha_hist_line]    64: ************* 413090
[M::ha_hist_line]    65: ************ 384390
[M::ha_hist_line]    66: *********** 354612
[M::ha_hist_line]    67: ********** 325811
[M::ha_hist_line]    68: ********* 295625
[M::ha_hist_line]    69: ******** 267797
[M::ha_hist_line]    70: ******** 243308
[M::ha_hist_line]    71: ******* 219014
[M::ha_hist_line]    72: ****** 200574
[M::ha_hist_line]    73: ****** 184990
[M::ha_hist_line]    74: ***** 171822
[M::ha_hist_line]    75: ***** 159544
[M::ha_hist_line]    76: ***** 149600
[M::ha_hist_line]    77: **** 141061
[M::ha_hist_line]    78: **** 135283
[M::ha_hist_line]    79: **** 131555
[M::ha_hist_line]    80: **** 129070
[M::ha_hist_line]    81: **** 126459
[M::ha_hist_line]    82: **** 126891
[M::ha_hist_line]    83: **** 125485
[M::ha_hist_line]    84: **** 125138
[M::ha_hist_line]    85: **** 123480
[M::ha_hist_line]    86: **** 122514
[M::ha_hist_line]    87: **** 122117
[M::ha_hist_line]    88: **** 117921
[M::ha_hist_line]    89: **** 117267
[M::ha_hist_line]    90: **** 114312
[M::ha_hist_line]    91: *** 110984
[M::ha_hist_line]    92: *** 106368
[M::ha_hist_line]    93: *** 103790
[M::ha_hist_line]    94: *** 100130
[M::ha_hist_line]    95: *** 96889
[M::ha_hist_line]    96: *** 93273
[M::ha_hist_line]    97: *** 89125
[M::ha_hist_line]    98: *** 85680
[M::ha_hist_line]    99: *** 82405
[M::ha_hist_line]   100: *** 80145
[M::ha_hist_line]   101: ** 77691
[M::ha_hist_line]   102: ** 76177
[M::ha_hist_line]   103: ** 73553
```

**Comment by chhylp123** [comment]
Yes, I agree. I guess in your current experiments, hifiasm automatically identified right and wrong hom coverage thresholds with `-s0.05` and `-s0.55`, respectively. So if you are using our latest release (v0.16.1-r375), I'd recommend you to try different values for `-s` always with `--hom-cov 57`. I guess `-s0.5` + `--hom-cov 57` might be able to produce two 2.8Gb haplotypes.

Based on the N50s, it seems your assemblies are not bad. I'm not definitely sure if 2.8Gb is too large for your samples, depending on how do you get the 2.5Gb estimated size. If 2.5Gb is anticipated by k-mers, it is usually too small. For example, the haploid genome size of the human genome is 3.1Gb while the k-mer estimated size is only 2.6Gb. For your cases, there might be two potential solutions: a) discarding all short contigs <1Mb (or other length thresholds) and seeing if the duplicated BUSCO scores can be decreased; b) manually checking duplicated genes identified BUSCO to make sure if they are real duplications. BUSCO score is not such accurate.

**Comment by pjm43** [comment]
You are correct, I did not specifically specify ```--homo-cov 57``` in the -s0.05.  I copied over the bin files from original default run and assumed that it would automatically use the same a homo-cov of 57.  I will rerun now in the -s0.05 folder, but this time I will specify specifically ```--homo-cov 57``` to see if that produces two equalized hap assemblies.

The estimated genome size is from flow cytometry work - but of course even those could be wrong.  We are going to recheck that now.  

Thanks again for your suggestions and help!

**Comment by chhylp123** [comment]
Probably you could specify `--homo-cov 57` with `-s0.55`. I guess when using `-s0.05`, the home coverage threshold has already been automatically set to 57.

**Comment by pjm43** [comment]
I apologize, I miss read your previous response.  You indicated that what I should try it -s0.5 (not -s0.05) + -hom-cov 57.  I will do that now and let you know how that come back.  Again, sorry about the confusion!
Jeff

**Comment by chhylp123** [comment]
: ) Look forwarding your results.

**Comment by pjm43** [comment]
Interesting results this morning.  I copied over the original bin files to new directories and then launched the same script (same -o) but this time with both ```-s 0.50``` and ```-hom-cov 57``` added to the script.  I did this for -s 0.4, 0.3, 0.2 and 0.1.  Weirdly, in all cases I get back the exact same assembly (regardless of -s value):  hap1 = 2.79 Gb and hap2 = 2.04 Gb.  I'm thinking I need to rerun hifiasm in it's entirety, instead of trying to reuse the bin file from the original run.  Do you agree?  Meaning I would run without any bin files present but add ```-s 0.50``` and ```-hom-cov 57``` to the command.  I realize this will take slightly longer to run.  FYI - the hifiasm version I'm running is: 0.16.1-r375.

Thanks again for your help!

**Comment by chhylp123** [comment]
Just make sure: are you using `-hom-cov 57` rather than `--hom-cov 57`? Have hifiasm actually run?

**Comment by pjm43** [comment]
So I'm now feeling not so smart!  I ran it with ```--homo-cov``` instead of ```--hom-cov```.  I'll rerun!  Sorry about wasting your time this morning!

**Comment by pjm43** [comment]
I've rerun with the correct flag this time: `-s 0.50` and `--hom-cov 57` (complete command: `hifiasm -o hifiasm_HiC -t 36 -s 0.50 --hom-cov 57 --h1 DTG-OmniC-223_R2_001.fastq.gz --h2 DTG-OmniC-223_R2_001.fastq.gz m54336U_210920_020003.hifi_reads.fastq.gz m54336U_211002_153817.hifi_reads.fastq.gz m54336U_211004_011130.hifi_reads.fastq.gz m64140_210926_030602.hifi_reads.fastq.gz m64140_210927_121816.hifi_reads.fastq.gz`). 

I still get high duplication rates with BUSCO, even when I reduce the `-s` flag and set `--hom-cov 57`.  I tried removing contigs below 1Mb, but the duplication rate for the haplotigs didn't change substantially (still are at about 21-25%).  Any additional thoughts about reducing duplication.  The genome size is also still a little higher than we expected.

Thanks

**Comment by chhylp123** [comment]
If the two haplotypes are balanced, the high BUSCO duplicated scores might be not a big deal. Some duplicated BUSCO genes are not the real duplications. I'd recommend you to check where are these duplicated genes to make sure if they are real.
